Blazor support to an ASP.NET Core app (Part 2) (#30799)
dotnet/AspNetCore.Docs: aspnetcore/blazor/components/integration.md

</html>
```

> For the `<link>` element in the preceding example, change `AspNetCoreApp` in the stylesheet's file name to match the app's namespace. For example, an app with the namespace `ContosoApp` uses the `ContosoApp.styles.css` stylesheet file name:

Reviewer: ```suggestion
For the `<link>` element in the preceding example, change `AspNetCoreApp` in the stylesheet's file name to match the app's project name. For example, an app project named `ContosoApp` uses the `ContosoApp.styles.css` stylesheet file name:
```

}
```

When the app is run, the `Counter` component is accessed at `/counter`.

### Enable interactive Auto and WebAssembly rendering

Follow the guidance in the [Add static server Razor component rendering](#add-static-server-razor-component-rendering) section.

> Auto render mode components render on the server first, then render on the client after the Blazor bundle has been downloaded and the Blazor runtime activates. The interactive WebAssembly render mode renders only renders components on the client after the Blazor bundle is downloaded and the Blazor runtime activates. Keep in mind that component code is ***not*** private using the Auto or WebAssembly render modes. For more information, see <xref:blazor/components/render-modes>.

Reviewer: Even with interactive WebAssembly rendering the components will by default still render first from the server as part of prerendering.

```suggestion
Components using the Auto render mode will initially use interactive server rendering, but then will switch to render on the client after the Blazor bundle has been downloaded and the Blazor runtime activates. Components using the WebAssembly render mode only render interactively on the client after the Blazor bundle is downloaded and the Blazor runtime activates. Keep in mind that when using the Auto or WebAssembly render modes, component code downloaded to the client is ***not*** private. For more information, see <xref:blazor/components/render-modes>.
```

* Pass the `-e|--empty` option.

From the donor Blazor Web App, copy the entire `.Client` project into the solution folder of the ASP.NET Core app.

> [!IMPORTANT]
> **Don't copy the `.Client` folder into the ASP.NET Core project's folder.** The best approach for organizing .NET solutions is to place each project of the solution into its own folder inside of a top-level solution folder. If a solution folder above the ASP.NET Core project's folder doesn't exist, create one. Next, copy the `.Client` project's folder from the donor Blazor Web App into the solution folder. The final project folder structure should have the following layout:
>
> * `AspNetCoreAppSolution` (top-level solution folder)
>   * `AspNetCoreApp` (original ASP.NET Core project)
>   * `AspNetCoreApp.Client` (`.Client` project folder from the donor Blazor Web App)
>
> For the ASP.NET Core solution file, you can leave it in the ASP.NET Core project's folder. Alternatively, you can move the solution file or create a new one in the top-level solution folder as long as the project references correctly point to the project files (`.csproj`) of the two projects in the solution folder.

> From the donor Blazor Web App, copy the `Components` folder into the ASP.NET Core app's project folder.
> 
> In the ASP.NET Core project's `Components/Pages` folder, delete the `Home` component (`Home.razor`).
> 

Reviewer: I don't see why these steps are necessary. Can they just be removed?

```suggestion
```

Author: The `Error` component has no value?

Reviewer: The Error component is useful for setting up an error page in your ASP.NET Core app when using the exception handling middleware, but it's not required.

If you named the donor Blazor Web App when you created the donor project the same as the ASP.NET Core app, the namespaces used by the donated assets match those in the ASP.NET Core app. You shouldn't need to take further steps to match namespaces. If you used a different namespace when creating the donor Blazor Web App project, you must adjust the namespaces across the donated assets to match if you intend to use the rest of this guidance exactly as presented. If the namespaces don't match, ***either*** adjust the namespaces before proceeding ***or*** adjust the namespaces as you follow the remaining guidance in this section.

Delete the donor Blazor Web App, as it has no further use in this process.

Add the `.Client` project to the solution:

* Visual Studio: Right-click the solution in **Solution Explorer** and select **Add** > **Existing Project**. Navigate to the `.Client` folder and select the project file (`.csproj`).

* .NET CLI: Use the [`dotnet sln add` command](/dotnet/core/tools/dotnet-sln#add) to add the `.Client` project to the solution.

> Add a project reference to the ASP.NET Core project for the client project:

Reviewer: ```suggestion
Add a project reference from the ASP.NET Core project to the client project:
```

  For interactive Auto rendering:

  ```csharp
  builder.Services.AddRazorComponents()
      .AddInteractiveServerComponents()
      .AddInteractiveWebAssemblyComponents();
  ```

  For only interactive WebAssembly rendering:

  ```csharp
  builder.Services.AddRazorComponents()
      .AddInteractiveWebAssemblyComponents();
  ```

* Add the interactive WebAssembly render mode (`AddInteractiveWebAssemblyRenderMode`) and additional assemblies for the `.Client` project where Razor components are mapped with `MapRazorComponents`.

  For interactive Auto rendering:

  ```csharp
  app.MapRazorComponents<App>()
      .AddInteractiveServerRenderMode()
      .AddInteractiveWebAssemblyRenderMode()
      .AddAdditionalAssemblies(typeof(AspNetCoreApp.Client._Imports).Assembly);
  ```

  For only interactive WebAssembly rendering:

  ```csharp
  app.MapRazorComponents<App>()
      .AddInteractiveWebAssemblyRenderMode()
      .AddAdditionalAssemblies(typeof(AspNetCoreApp.Client._Imports).Assembly);
  ```

  In the preceding examples, change `AspNetCoreApp.Client` to match the `.Client` project's namespace.

> Add a `Components` folder and a `Components/Pages` folder to the `.Client` project.
> 
> Add the following `Counter` component (`Components/Pages/Counter.razor`) to the `.Client` project:

Reviewer: When the project is Blazor specific, like in a Blazor client project, we decide to drop the `Components` folder.

```suggestion
Add a `Pages` folder to the `.Client` project.

Add the following `Counter` component (`Pages/Counter.razor`) to the `.Client` project:
```

> Add the following `Counter` component (`Components/Pages/Counter.razor`) to the `.Client` project:

Reviewer: If the user already added a Counter component to the server project, then they may end up with two Counter components that have a routing conflict. If they already have a Counter can we tell them to just move it to the client, otherwise add one?

Author: It instructs them not to add the sample code in the earlier step.

Author: Lines 238 and 244.

Author: Do you still want me to add a line with that sentiment :point_up:? ... Move it if you have it.

Author: ***Nevermind*** ... I forgot what we were doing with all of the churn. I'll add a remark on it :point_up:.
